File: feature-timesheet/src/main/java/fr/o80/sample/timesheet/presentation/ui/TimesheetAdapter.java
package fr.o80.sample.timesheet.presentation.ui;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import fr.o80.sample.timesheet.R;
import fr.o80.sample.timesheet.R2;
import fr.o80.sample.timesheet.data.entity.TimeEntry;

/**
 * @author Olivier Perez
 */
public class TimesheetAdapter extends RecyclerView.Adapter<TimesheetAdapter.EntryViewHolder> {

    private List<TimeEntry> items;

    @Override
    public EntryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_timesheet, parent);
        return new EntryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(EntryViewHolder holder, int position) {
        holder.bind(items.get(position));
    }

    @Override
    public int getItemViewType(int position) {
        return super.getItemViewType(position);
    }

    @Override
    public int getItemCount() {
        return items == null ? 1 : items.size();
    }

    public void setItems(List<TimeEntry> entries) {
        items = entries;
        notifyItemRangeInserted(0, entries.size());
    }

    protected class EntryViewHolder extends RecyclerView.ViewHolder {

        @BindView(R2.id.project_name)
        protected TextView projectName;

        @BindView(R2.id.project_code)
        protected TextView projectCode;

        public EntryViewHolder(View itemView) {
            super(itemView);
        }

        public void bind(TimeEntry timeEntry) {
            projectName.setText(timeEntry.project());
            projectCode.setText(timeEntry.code());
        }
    }
}

File: feature-timesheet/src/main/java/fr/o80/sample/timesheet/presentation/ui/TimesheetAdapter.kt
package fr.o80.sample.timesheet.presentation.ui

import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import fr.o80.sample.timesheet.R
import fr.o80.sample.timesheet.usecase.model.EntryCallback
import fr.o80.sample.timesheet.usecase.model.EntryViewModel

/**
 * @author Olivier Perez
 */
class TimesheetAdapter(
        val onClick: (EntryViewModel) -> Unit,
        val onTimeAdded: (EntryViewModel) -> Unit,
        val onLongTimeAdded: (EntryViewModel) -> Boolean,
        val onTimeRemoved: (EntryViewModel) -> Unit,
        val onLongTimeRemoved: (EntryViewModel) -> Boolean
) : ListAdapter<EntryViewModel, TimesheetAdapter.EntryViewHolder>(EntryCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EntryViewHolder =
            EntryViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_timesheet, parent, false))

    override fun onBindViewHolder(holder: EntryViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class EntryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val projectName: TextView = itemView.findViewById(R.id.projectName)
        private val projectCode: TextView = itemView.findViewById(R.id.projectCode)
        private val time: TextView = itemView.findViewById(R.id.time)
        private val addTime: TextView = itemView.findViewById(R.id.plusButton)
        private val removeTime: TextView = itemView.findViewById(R.id.minusButton)

        fun bind(timeEntry: EntryViewModel) {
            with(timeEntry) {
                projectName.text = label
                projectCode.text = code
                time.text = itemView.context.getString(R.string.hours, hours)
            }
            itemView.setOnClickListener { onClick(timeEntry) }
            addTime.setOnClickListener { onTimeAdded(timeEntry) }
            addTime.setOnLongClickListener { onLongTimeAdded(timeEntry) }
            removeTime.setOnClickListener { onTimeRemoved(timeEntry) }
            removeTime.setOnLongClickListener { onLongTimeRemoved(timeEntry) }
        }
    }
}
